### PR TITLE
Feature: Improve API helpers and error handling callbacks

### PR DESCRIPTION
**Pull request by seangwright**
This PR changes how the `fetchResponse` helper works by returning the `Response` object instead of mapping it to `Unit`. This lets consumers use all the `Response` state to make additional logic decisions.

It also makes all the error handlers callbacks for async operations optionally async.

Finally, some missing tests were added for `ResultAsync.try`, and the `fetchResponse`/`fetchJsonResponse` helpers. There are still more test cases that need to be added for async error handlers, but there are no regressions with existing functionality, so the additional tests can wait for a later PR.

Because of the change to `fetchResponse`, this will result in a major version bump to `2.0.0`.

### CODE REVIEW

File: src/api.ts [hunk 1]
```diff
@@ -27,18 +28,19 @@ export function fetchJsonResponse<TValue, TError = string>(
...
- * The response body is mapped to Unit. Use this for No Content responses.
+ * The raw Response object is returned as the Result success value.
  * @param request A Promise<Response> generated by a fetch() request
- * @param errorHandler Handles connection and Http status code errors
+ * @param errorHandler Handles connection and Http status code errors. Can be asynchronous to enable evaluation of the Response body.
  * @returns ResultAsync representing the success/failure of the request
  */
 export function fetchResponse<TError = string>(
   request: Promise<Response>,
-  errorHandler: (error: unknown) => Some<TError>
```
Comment (seangwright) [inline src/api.ts:37]: Not sure why this wasn't retyped as `FunctionOfTtoK` originally 🤔. Must have missed it.

File: src/resultAsync.ts [hunk 2]
```diff
@@ -99,34 +105,53 @@ export class ResultAsync<TValue = Unit, TError = string> {
...
+          .catch(unwrapHandledError(errorHandler))
       );
     } catch (error: unknown) {
-      return new ResultAsync(
-        Promise.resolve(Result.failure(errorHandler(error)))
-      );
+      return new ResultAsync(unwrapHandledError(errorHandler)(error));
+    }
+
+    function unwrapHandledError(
```
Comment (seangwright) [inline src/resultAsync.ts:140]: There might be a benefit to making this a function private to the module but not something that closes over the state of `try`.

Comment (seangwright) [reply src/resultAsync.ts]: Moved this to the bottom of the module and re-used it in other methods 🍻

File: test/api/fetchResponse.spec.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
+import { fetchResponse } from '@/src/api';
+import { Response } from 'cross-fetch';
```
Comment (seangwright) [inline test/api/fetchResponse.spec.ts:2]: `Response` doesn't exist in the node environment used in GitHub actions, so even though the test passes without this locally it fails in CI.

`cross-fetch` is acting a pre-built mock for the behavior of the browser `Response` object.

As far as the `fetchResponse`/`fetchJsonResponse` implementations go, the `Response` type just needs an `ok` property - nothing else is accessed/checked.

Comment (seangwright) [reply test/api/fetchResponse.spec.ts]: Just bumped the node version, which was 16, to 18 to give support to fetch/Response

**Review by xShivan — APPROVED** [review]

**Review by xShivan — DISMISSED** [review]

**Comment by xShivan** [comment]
Node.js version in README.md should be updated too.

**Comment by seangwright** [comment]
Good catch! I'll also want to update the stats of the file sizes since they were modified.

File: README.md [hunk 4]
```diff
@@ -4,12 +4,24 @@
 
 A TypeScript implementation of the C# library [CSharpFunctionalExtensions](https://github.com/vkhorikov/CSharpFunctionalExtensions), including synchronous and asynchronous `Maybe` and `Result` monads.
 
-## Further Reading
+## Community
+
+### Related Projects
+
+- [NestJS typescript-functional-extensions utilities](https://github.com/startupdevhouse/typescript-functional-extensions-nestjs) (A library of utilities for working with `typescript-functional-extensions` in [NestJS](https://nestjs.com/) projects)
```
Comment (KyleMcMaster) [inline README.md:11]: Nice shoutout! 👍

**Review by KyleMcMaster — APPROVED** [review]
:shipit:

**Pull request merged**